paperscraper/lib.py

```diff
@@ -617,6 +720,170 @@ async def google2s2(
     return paths
 
 
+async def a_gsearch_papers(  # noqa: C901, PLR0915
+    query: str,
+    limit=10,
+    pdir=os.curdir,
+    _paths: dict[str | os.PathLike, dict[str, Any]] | None = None,
+    _offset: int = 0,
+    _limit: int = 20,
+    logger: logging.Logger | None = None,
+    year: str | None = None,
+    verbose: bool = False,
+    scraper: Scraper | None = None,
+    batch_size: int = 10,
+) -> dict[str, dict[str, Any]]:
+    if not os.path.exists(pdir):
+        os.mkdir(pdir)
+    if logger is None:
+        logger = logging.getLogger("paper-scraper")
+        logger.setLevel(logging.ERROR)
+        if verbose:
+            logger.setLevel(logging.DEBUG)
+            ch = logging.StreamHandler()
+            ch.setFormatter(CustomFormatter())
+            logger.addHandler(ch)
+    endpoint = "https://serpapi.com/search.json"
+    params = {
+        "q": query,
+        "api_key": os.environ["SERPAPI_API_KEY"],
+        "engine": "google_scholar",
+        "num": _limit,
+        "start": _offset,
+        # TODO - add offset and limit here  # noqa: TD004
+    }
+
+    if year is not None:
+        # need to really make sure year is correct
+        year = year.strip()
+        if "-" in year:
+            # make sure start/end are valid
+            try:
+                start, end = year.split("-")
+                if int(start) <= int(end):
+                    params["as_ylo"] = start
+                    params["as_yhi"] = end
+            except ValueError:
+                pass
+        else:
+            with contextlib.suppress(ValueError):
+                params["as_ylo"] = year
+                params["as_yhi"] = year
+        if "as_ylo" not in params:
+            logger.warning(f"Could not parse year {year}")
+
+    paths = _paths or {}
+    scraper = scraper or default_scraper()
+    ssheader = get_header()
+    # add key to headers
+
+    async with ThrottledClientSession(
+        headers=ssheader,
+        rate_limit=30 / 60,
+    ) as gs_session:
+        async with gs_session.get(
+            url=endpoint,
+            params=params,
+        ) as response:
+            if not response.ok:
+                raise RuntimeError(
+                    f"Error searching papers: {response.status} {response.reason} {await response.text()}"  # noqa: E501
+                )
+            data = await response.json()
+
+        # session for cross-ref
+        async with ThrottledClientSession(
+            headers=ssheader,
+            rate_limit=50,
+        ) as cr_session:
```

Comment:
Can we dedent the second session, so it can't confuse one to have nested sessions?

Reply:
We can just use one session I think - will change to that